lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,969 @@
+      std::is_same<A, parser::EndMpSubprogramStmt>,
+      std::is_same<A, parser::EndProgramStmt>,
+      std::is_same<A, parser::EndSubroutineStmt>>;
+}
+template<>
+constexpr bool IsLegalBranchTarget(
+    const parser::Statement<parser::ActionStmt> &actionStmt) {
+  if (!isStrictF18) {
+    return true;
+  } else {
+    // XXX: do we care to flag these as errors? If we want strict F18, these
+    // statements should not even be present
+    return !(
+        std::holds_alternative<common::Indirection<parser::ArithmeticIfStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::AssignStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::AssignedGotoStmt>>(
+            actionStmt.statement.u) ||
+        std::holds_alternative<common::Indirection<parser::PauseStmt>>(
+            actionStmt.statement.u));
+  }
+}
+
+template<typename A>
+constexpr TargetStmtType ConsTrgtFlags(const parser::Statement<A> &statement) {
+  TargetStmtType targetStmtType{};
+  if (IsLegalDoTerm(statement)) {
+    targetStmtType.set(TargetStatementEnum::Do);
+  }
+  if (IsLegalBranchTarget(statement)) {
+    targetStmtType.set(TargetStatementEnum::Branch);
+  }
+  if (IsFormat(statement)) {
+    targetStmtType.set(TargetStatementEnum::Format);
+  }
+  return targetStmtType;
+}
+
+/// \brief \p opt1 and \p opt2 must be either present and identical or absent
+/// \param a  an optional construct-name (opening statement)
+/// \param b  an optional construct-name (ending statement)
+template<typename A> inline bool BothEqOrNone(const A &a, const A &b) {
+  return (a.has_value() == b.has_value())
+      ? (a.has_value() ? (a.value().ToString() == b.value().ToString()) : true)
+      : false;
+}
+
+/// \brief \p opt1 must either be absent or identical to \p b
+/// \param a  an optional construct-name for an optional constraint
+/// \param b  an optional construct-name (opening statement)
+template<typename A> inline bool PresentAndEq(const A &a, const A &b) {
+  return (!a.has_value()) ||
+      (b.has_value() && (a.value().ToString() == b.value().ToString()));
+}
+
+/// \brief Iterates over parse tree, creates the analysis result
+/// As a side-effect checks the constraints for the usages of
+/// <i>construct-name</i>.
+struct ParseTreeAnalyzer {
+public:
+  struct UnitAnalysis {
+  public:
+    SourceStmtList doStmtSources_;  ///< bases of label-do-stmts
+    SourceStmtList
+        formatStmtSources_;  ///< bases of all other stmts with labels
+    SourceStmtList otherStmtSources_;  ///< bases of all other stmts with labels
+    TargetStmtMap targetStmts_;  ///< unique map of labels to stmt info
+    std::vector<ScopeProxy> scopeModel_;  ///< scope stack model
+
+    UnitAnalysis() { scopeModel_.push_back(0); }
+    UnitAnalysis(UnitAnalysis &&) = default;
+    ~UnitAnalysis() = default;
+    UnitAnalysis(const UnitAnalysis &) = delete;
```

Comment:
Deletion of the copy constructor is implied by the C++ programming language in the presence of a move constructor.

Reply:
expanded to user-defined move constructor